src/motor-operations.cc

```diff
@@ -84,6 +85,95 @@ static char test_acceleration_ok(float acceleration) {
 }
 #endif
 
+struct HistorySlot {
+  int steps[BEAGLEG_NUM_MOTORS];
+  int max_axis_steps;
+};
+
+typedef RingDeque<struct HistorySlot, QUEUE_LEN + 1> SlotHistory;
+typedef FixedArray<unsigned,BEAGLEG_NUM_MOTORS> SlotPartial;
+
+// This class is delegated to handle the update of the absolute
+// position in steps of each motor and it should be callbacked whenever this
+// operations has to be performed.
+class MotionQueueMotorOperations::StatusListener
+                                  : public MotionQueue::PositionChangeCallback {
+public:
+  StatusListener(MotionQueue *backend, MotorCoordSystem *current_position_steps)
+                : backend_(backend),
+                current_position_steps_(current_position_steps) {}
+  virtual void UpdateAbsolutePosition();
+  void PushInHistory(const LinearSegmentSteps &param,
+                       const int defining_axis_steps) {
+    if (execution_history_.size() == QUEUE_LEN) {
+      Log_error("The queue slots have been executed faster than beagleg "
+                "pushing new ones.");
+      exit(1);
+    }
+    struct HistorySlot *just_uploaded = execution_history_.append();
+    for (int i = 0; i < MOTION_MOTOR_COUNT; ++i) {
+      just_uploaded->steps[i] = param.steps[i];
+    }
+    just_uploaded->max_axis_steps = defining_axis_steps;
+  }
+
+private:
+  MotionQueue *const backend_;
+  SlotHistory execution_history_;
+  SlotPartial last_partial;
+  MotorCoordSystem *const current_position_steps_;
+};
+
+void MotionQueueMotorOperations::StatusListener::UpdateAbsolutePosition() {
+  int i;
+  struct QueueStatus status_register;
+  backend_->GetStatus(&status_register);
+  if (execution_history_.size() == 0) return;
+  // Pru_queue_size represent the real pru queue size, so we shrink the history
+  // queue until it's exactly the same as the pru one.
+  const unsigned pru_queue_size
+    = ((execution_history_.back_index() - 1) % (QUEUE_LEN + 1)
```

Comment:
This breaks the encapsulation of the execution_history_ ... you are essentially doing things here that you only can do with the knowledge how the execution history is dealt with internally.

Anyway, this part is extremely hard to understand, what is happening here ?